python/cuml/ts/arima.pyx

```diff
@@ -0,0 +1,814 @@
+        Time series data.
+    order : Tuple[int, int, int]
+            The ARIMA order (p, d, q)
+    mu0 : array-like
+          Array of trend-parameter estimates. Only used if `d>0`.
+    ar_params0 : List of arrays
+                 List of AR parameter-arrays, one array per series, each series has `p` parameters.
+    ma_params0 : List of arrays
+                 List of MA parameter-arrays, one array per series, each series has `q` parameters.
+    opt_disp : int
+               Fit diagnostic level (for L-BFGS solver):
+               * `-1` for no output,
+               * `0<n<100` for output every `n` steps
+               * `n>100` for more detailed output
+    h        : float
+               Finite-differencing step size. The gradient is computed using second-order differencing:
+                   f(x+h) - f(x - h)
+               g = ----------------- + O(h^2)
+                         2 * h
+
+    Returns:
+    --------
+    model : ARIMAModel
+            The ARIMA model with the best fit parameters.
+
+    """
+    p, d, q = order
+    num_parameters = d + p + q
+
+    d_y, d_y_ptr, num_samples, num_batches, dtype = input_to_dev_array(y, check_dtype=np.float64)
+
+    handle = cuml.common.handle.Handle()
+
+    def f(x: np.ndarray) -> np.ndarray:
+        """The (batched) energy functional returning the negative loglikelihood (for each series)."""
+
+        # Recall: Maximimize LL means minimize negative
+        n_llf = -(ll_f(num_batches, num_samples, order, d_y, x, trans=True, handle=handle))
+        return n_llf/(num_samples-1)
+
+
+    # optimized finite differencing gradient for batches
+    def gf(x):
+        """The gradient of the (batched) energy functional."""
+        # Recall: We maximize LL by minimizing -LL
+        n_gllf = -ll_gf(num_batches, num_samples, num_parameters, order, d_y, x, h, trans=True, handle=handle)
+        return n_gllf/(num_samples-1)
+
+    x0 = pack(p, d, q, num_batches, mu0, ar_params0, ma_params0)
+    x0 = _batch_invtrans(p, d, q, num_batches, x0, handle)
+
+    # check initial parameter sanity
+    if ((np.isnan(x0).any()) or (np.isinf(x0).any())):
+            raise FloatingPointError("Initial condition 'x0' has NaN or Inf.")
+
+
+    # Optimize parameters by minimizing log likelihood.
+    x, niter, flags = batched_fmin_lbfgs_b(f, x0, num_batches, gf,
+                                           iprint=opt_disp, factr=1000)
+
+    # TODO: Better Handle non-zero `flag` array values: 0 -> ok, 1,2 -> optimizer had trouble
+    if (flags != 0).any():
+        print("WARNING(`fit()`): Some batch members had optimizer problems.")
+
+    Tx = _batch_trans(p, d, q, num_batches, x, handle)
+    mu, ar, ma = unpack(p, d, q, num_batches, Tx)
+
+    fit_model = ARIMAModel(num_batches*[order], mu, ar, ma, y)
+    fit_model.niter = niter
+    fit_model.d_y = (d_y, d_y_ptr)
+    return fit_model
+
+
+
+def grid_search(y_b, d=1, max_p=3, max_q=3, method="bic"):
```

Comment:
We probably want to think soon after the initial version is in how to couple this with the dask grid search work that is being done

Reply:
Obviously this is a basic implementation. I wanted to have it because one of the reference libraries @jeaton32 provided (`pmdarima`) has an "auto-arima" function which is just grid search.